### ISSUE TITLE
read-backup-data parameter not working as expected

### ISSUE THREAD

**Issue by arnaud-soulard** [hazelcast/hazelcast#11450]
The context is the following:

1. Hazelcast version 3.8.6
2. I want to use 2 members
3. I am using the provided console.sh, to start com.hazelcast.console.ConsoleApp
4. Java version:
openjdk version "1.8.0_141"
OpenJDK Runtime Environment (build 1.8.0_141-b16)
OpenJDK 64-Bit Server VM (build 25.141-b16, mixed mode)
5. OS: CentOS Linux release 7.3.1611 (Core)
uname -a: Linux mfpz-1 3.10.0-514.26.2.el7.x86_64 #1 SMP Tue Jul 4 15:04:05 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
6-7) Hereafter is the scenario:
I start 2 nodes.
```
Members [2] {
	Member [10.86.84.178]:5701 - 1277609f-dd3e-4fb4-a78c-a69d3f46b360
	Member [10.86.86.50]:5701 - 19006ecf-ea89-4219-8d5f-840e3db088f9 this
}
```

The partitioning should be the following:

![backuppartitions](https://user-images.githubusercontent.com/10761110/30853391-f709aa74-a2ae-11e7-962a-211328eebf7e.png)

I am working on the first node.
In the configuration file, I set `read-backup-data` to true.

On the first node, I put a key.
Then, I run `m.size` to check it has been set.
I did not specify the property `hazelcast.max.no.heartbeat.seconds`, it is set to 300 seconds by default.
If I disable node2 (ifdown on network interface), I expect the data to be read from node1, since `read-backup-data` is set to true.
In fact, `m.size` hangs, and do not return until `hazelcast.max.no.heartbeat.seconds` is reached.

In other experiments, I set `hazelcast.max.no.heartbeat.seconds` to 4 seconds. But I need to have a quicker response from the system, ideally without repartitioning, since it should not be needed in this case (we have all informations on node1!)

Here is the configuration file:
[hazelcast.xml.txt](https://github.com/hazelcast/hazelcast/files/1332662/hazelcast.xml.txt)

**Comment by mmedenjak** [comment]
Hi @arnaud-soulard !

The `read-backup-data` only applies to reading map entries by key, not to any other map action such as `size()`. Even though the map size could be determined here since the member contains all partitions (either it is the owner or backup), this is not the case when you have more than 2 members in the cluster.
In that case, a single member cannot determine the size of the map and must thus wait for all members to return the owned partitions size or timeout.
So I would count this as an optimisation, not necessarily as a defect. 
Or does reading from the map hang when node2 is disabled?

**Comment by metanet** [comment]
Please also note that the partition distribution in the figure is shown for the sake of simplicity. Partitions do not have to be sequentially distributed to each node.

**Comment by arnaud-soulard** [comment]
To answer the question: "Does reading from the map hang when node2 is disabled?"
No, you're right, a "m.get" does not hang.
I used to work with "m.size", to check that all members (when using more than 2 members) know the same entries. When enabling/disabling members with "hazelcast.max.no.heartbeat.seconds" set to 4 seconds, a Merge is initiated, and if not terminated, a "m.size" on all members could give different results.

**Comment by mmedenjak** [comment]
That's expected behaviour - `IMap#size` does not check the local state only and is not affected by the `read-backup-data` parameter - it always sends an operation to all known members to collect their owned partition sizes. 
The method can return different results if there is a network partition and some partitions are completely isolated from a cluster (both partition owner and backup replicas).

**Comment by arnaud-soulard** [comment]
Since the "m.get" command is working when I use console.sh ("com.hazelcast.console.ConsoleApp") when node2 is disabled, I wanted to use the nodejs client for the same use case. But with this nodejs client, it hangs as soon as the node2 is disabled. The client certainly execute other commands, prior to the get.

**Comment by mmedenjak** [comment]
I'm a bit confused. First we concluded:
> No, you're right, a "m.get" does not hang.
Then:
> But with this nodejs client, it hangs as soon as the node2 is disabled.
And we haven't mentioned nodejs previously in the issue. I would summarise this as:
1) `map#size` is working as expected, although it blocks
2) `map#get` blocks only on the nodejs client

Is this correct?

**Comment by arnaud-soulard** [comment]
exactly!

**Comment by mmedenjak** [comment]
I guess it hangs regardless of the `read-backup-data` parameter then?

**Comment by arnaud-soulard** [comment]
Exactly again!
I'm still doing some tests (without nodejs client, inside console.sh) when a node becomes unavailable. Finally, m.get works, but if I want to insert a new key (m.put) when the second node is disabled, the command (m.put) hangs.

**Comment by mmedenjak** [comment]
The fact that a put operation hangs is also expected. The `read-backup-data` parameter only allows reads if a node contains a backup of a partition. Mutation is not allowed on a backup and if the second node is unreachable, put operations are expected to hang.

One more thing - the `read-backup-data` only affects get operations when invoked on the hazelcast member. Hazelcast clients will always try and read from the partition owner which in your case is not reachable. So it is also expected that `map#get` only blocks on a node.js client.

To summarise, if you make the second node unreachable, the `read-backup-data` will only allow get operations to read from backups and only when calling `map#get` on a hazelcast member directly. 
I am closing this issue since all of the behaviour is expected. If you have any other problems, please open a new issue.
